plenum/recorder/src/replayer.py

```diff
@@ -0,0 +1,178 @@
+    # Always add node recorder first and then client recorder
+    cr.add_recorders(node_recorder, client_recorder)
+    cr.combine_recorders()
+    return cr
+
+
+def prepare_node_for_replay_and_replay(looper, replaying_node,
+                                       node_recorder, client_recorder,
+                                       start_times):
+    cr = get_combined_recorder(replaying_node, node_recorder, client_recorder)
+    cr.start_times = start_times
+    patch_replaying_node(replaying_node, node_recorder, start_times)
+    return replay_patched_node(looper, replaying_node, node_recorder, cr)
+
+
+def patch_replaying_node(replaying_node, node_recorder, start_times):
+    patch_sent_prepreapres(replaying_node, node_recorder)
+    patch_replaying_node_for_time(replaying_node, start_times)
+
+
+def patch_replaying_node_for_time(replaying_node, start_times):
+    node_1st_start_time = start_times[0][0]
+    replaying_node._time_diff = get_utc_epoch() - node_1st_start_time
+
+
+def replay_patched_node(looper, replaying_node, node_recorder, cr):
+    node_run_no = 0
+    looper.add(replaying_node)
+    cr.start_playing()
+    next_stop_at = time.perf_counter() + (cr.start_times[node_run_no][1] -
+                                          cr.start_times[node_run_no][0])
+    while cr.is_playing:
+        vals = cr.get_next()
+        if next_stop_at is not None and time.perf_counter() >= next_stop_at:
+            node_run_no += 1
+            if node_run_no < len(cr.start_times):
+                # The node stopped here
+                sleep_for = cr.start_times[node_run_no][0] - cr.start_times[node_run_no - 1][1]
+                replaying_node.stop()
+                looper.removeProdable(replaying_node)
+                # Create new node since node is destroyed on stop
+                replaying_node = replaying_node.__class__(replaying_node.name,
+                                                          config_helper=replaying_node.config_helper,
+                                                          ha=replaying_node.nodestack.ha,
+                                                          cliha=replaying_node.clientstack.ha)
+                patch_replaying_node(replaying_node, node_recorder, cr.start_times)
+                print('sleeping for {} to simulate node stop'.format(sleep_for))
+                time.sleep(sleep_for)
+                after = cr.start_times[node_run_no][1] - cr.start_times[node_run_no][0]
+                next_stop_at = time.perf_counter() + after
+                print('Next stop after {}'.format(after))
+                looper.add(replaying_node)
+            else:
+                next_stop_at = None
+
+        if not vals:
+            # looper.runFor(.001)
+            continue
+
+        n_msgs, c_msgs = vals
+        if n_msgs:
+            for inc in n_msgs:
+                if Recorder.is_incoming(inc):
+                    replaying_node.nodestack._verifyAndAppend(to_bytes(inc[1]),
```

Comment:
What does indices [0], [1], [2] mean here?

Reply:
Will clarify in code